### ISSUE TITLE
Last frame of video jumps when switching to new video with setKeepContentOnPlayerReset(true)

### ISSUE THREAD

**Issue by CleverSJX** [google/ExoPlayer#9835]
The 16:9 video played in portrait mode is displayed in the center by default. In order to prevent the video from switching to a black screen, I called ```playerView.setKeepContentOnPlayerReset(true);```, but when switching the video, the last frame of the video appears at the top, or is pulled up by the full screen

[Recorded video](https://s.pointshow.net/transcode/normal/QPnNC1641350941063.mp4)

```
    private void play(int position) {
        initPlayer();
        Log.i(TAG, "play position: " + position);
        MediaItem mediaItem = mediaItemList.get(position);
        assert mediaItem.localConfiguration != null;
        if (coverView != null) {
            coverView.setVisibility(View.VISIBLE);
            Glide.with(this).load(mediaItem.localConfiguration.tag).into(coverView);
        }
        exoPlayer.setMediaItem(mediaItem);
        exoPlayer.prepare();
    }

    private void initData() {
        MediaItem firstItem = new MediaItem.Builder()
                .setUri(VIDEO_PATH + "1.mp4")
                .setTag(VIDEO_PATH + "1.jpg")
                .build();
        MediaItem secondItem = new MediaItem.Builder()
                .setUri(VIDEO_PATH + "2.mp4")
                .setTag(VIDEO_PATH + "2.jpg")
                .build();
        mediaItemList.add(firstItem);
        mediaItemList.add(secondItem);
    }

    private void initPlayer() {
        if (exoPlayer != null) {
            exoPlayer.release();
        }
        exoPlayer = new ExoPlayer
                .Builder(this)
                .build();
        playerView.setResizeMode(AspectRatioFrameLayout.RESIZE_MODE_FIT);
        playerView.setPlayer(exoPlayer);
        playerView.setKeepContentOnPlayerReset(true);

        exoPlayer.addAnalyticsListener(new EventLogger(null));
        exoPlayer.addListener(new MyListener());
        exoPlayer.setPlayWhenReady(true);
        exoPlayer.setRepeatMode(Player.REPEAT_MODE_OFF);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (exoPlayer != null) {
            exoPlayer.release();
        }
    }

    private class MyListener implements Player.Listener {

        @Override
        public void onPlayerError(@NonNull PlaybackException error) {
            Log.e(TAG, "onPlayerError: ", error);
        }

        @Override
        public void onMediaItemTransition(@Nullable MediaItem mediaItem, int reason) {
            Log.i(TAG, "onMediaItemTransition: mediaItem: " + mediaItem.localConfiguration.uri);
        }

        @Override
        public void onRenderedFirstFrame() {
            Log.i(TAG, "onRenderedFirstFrame: " + "gone view ");
            if (coverView != null) {
                coverView.setVisibility(View.GONE);
            }
        }

        @Override
        public void onPlaybackStateChanged(int playbackState) {
//            Log.i(TAG, "onPlaybackStateChanged: playbackState: " + exoPlayer.getCurrentMediaItem().localConfiguration.uri);
            switch (playbackState) {
                case Player.STATE_IDLE:
                    Log.i(TAG, "STATE_IDLE");
                    break;
                case Player.STATE_BUFFERING:
                    Log.i(TAG, "STATE_BUFFERING");
                    break;
                case Player.STATE_READY:
                    Log.i(TAG, "STATE_READY");
                    break;
                case Player.STATE_ENDED:
                    Log.i(TAG, "STATE_ENDED");
                    complete();
                    break;
            }
        }

        @Override
        public void onIsPlayingChanged(boolean isPlaying) {
            Log.i(TAG, "onIsPlayingChanged: " + isPlaying);
        }
    }

    private void complete() {
        if (currentPosition >= mediaItemList.size() - 1) {
            currentPosition = 0;
        } else {
            currentPosition++;
        }
        play(currentPosition);
    }
```

- ExoPlayer version number: 2.16.1
- Android version: all
- Android device:all

**Comment by ojw28** [comment]
There is no need to release the player and instantiate a new one whenever you want to play something else. Just call `stop()` on the existing player, then pass the new item to play via `setMediaItem(...)` and call `prepare(...)`.

Doing it this way is more efficient, and avoids needing to reconnect the view to a different player instance. Please try this and let us know whether this resolves your issue (you probably still need to call `setKeepContentOnPlayerReset`, but only once during initial configuration).

It probably is also a bug that the way you're doing it doesn't work quite right, so we can look at fixing that as a low priority issue. But either way, you're better off adapting your code as described above. Please let us know how you get on!

**Comment by CleverSJX** [comment]
> There is no need to release the player and instantiate a new one whenever you want to play something else. Just call `stop()` on the existing player, then pass the new item to play via `setMediaItem(...)` and call `prepare(...)`.
> 
> Doing it this way is more efficient, and avoids needing to reconnect the view to a different player instance. Please try this and let us know whether this resolves your issue (you probably still need to call `setKeepContentOnPlayerReset`, but only once during initial configuration).
> 
> It probably is also a bug that the way you're doing it doesn't work quite right, so we can look at fixing that as a low priority issue. But either way, you're better off adapting your code as described above. Please let us know how you get on!

Thank you very much, according to the usage method you said, the problem is solved, and the position of the last frame is normal when the video is switched.

**Comment by CleverSJX** [comment]
> > There is no need to release the player and instantiate a new one whenever you want to play something else. Just call `stop()` on the existing player, then pass the new item to play via `setMediaItem(...)` and call `prepare(...)`.
> > Doing it this way is more efficient, and avoids needing to reconnect the view to a different player instance. Please try this and let us know whether this resolves your issue (you probably still need to call `setKeepContentOnPlayerReset`, but only once during initial configuration).
> > It probably is also a bug that the way you're doing it doesn't work quite right, so we can look at fixing that as a low priority issue. But either way, you're better off adapting your code as described above. Please let us know how you get on!
> 

Thank you very much, according to the usage method you said, the problem is solved, and the position of the last frame is normal when the video is switched.
```

    private void play(int position) {
        exoPlayer.stop();
        Log.i(TAG, "play position: " + position);
        MediaItem mediaItem = mediaItemList.get(position);
        assert mediaItem.localConfiguration != null;
        if (coverView != null) {
            coverView.setVisibility(View.VISIBLE);
            Glide.with(this).load(mediaItem.localConfiguration.tag).into(coverView);
        }
        exoPlayer.setMediaItem(mediaItem);
        exoPlayer.prepare();
    }

    private void initData() {
        // Build the media items.
        MediaItem firstItem = new MediaItem.Builder()
                .setUri(VIDEO_PATH + "1.mp4")
                .setTag(VIDEO_PATH + "1.jpg")
                .build();
        MediaItem secondItem = new MediaItem.Builder()
                .setUri(VIDEO_PATH + "2.mp4")
                .setTag(VIDEO_PATH + "2.jpg")
                .build();
        mediaItemList.add(firstItem);
        mediaItemList.add(secondItem);
    }

    private void initPlayer() {
        exoPlayer = new ExoPlayer
                .Builder(this)
                .build();
        playerView.setPlayer(exoPlayer);
        playerView.setKeepContentOnPlayerReset(true);
        exoPlayer.addAnalyticsListener(new EventLogger(null));
        exoPlayer.addListener(new MyListener());
        exoPlayer.setPlayWhenReady(true);
        exoPlayer.setRepeatMode(Player.REPEAT_MODE_OFF);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (exoPlayer != null) {
            exoPlayer.release();
        }
    }

    private class MyListener implements Player.Listener {

        @Override
        public void onPlayerError(@NonNull PlaybackException error) {
            Log.e(TAG, "onPlayerError: ", error);
        }

        @Override
        public void onMediaItemTransition(@Nullable MediaItem mediaItem, int reason) {
            Log.i(TAG, "onMediaItemTransition: mediaItem: " + mediaItem.localConfiguration.uri);
        }

        @Override
        public void onRenderedFirstFrame() {
            Log.i(TAG, "onRenderedFirstFrame: " + "gone view ");
            if (coverView != null) {
                coverView.setVisibility(View.GONE);
            }
        }

        @Override
        public void onPlaybackStateChanged(int playbackState) {
//            Log.i(TAG, "onPlaybackStateChanged: playbackState: " + exoPlayer.getCurrentMediaItem().localConfiguration.uri);
            switch (playbackState) {
                case Player.STATE_IDLE:
                    Log.i(TAG, "STATE_IDLE");
                    break;
                case Player.STATE_BUFFERING:
                    Log.i(TAG, "STATE_BUFFERING");
                    break;
                case Player.STATE_READY:
                    Log.i(TAG, "STATE_READY");
                    break;
                case Player.STATE_ENDED:
                    Log.i(TAG, "STATE_ENDED");
                    complete();
                    break;
            }
        }

        @Override
        public void onIsPlayingChanged(boolean isPlaying) {
            Log.i(TAG, "onIsPlayingChanged: " + isPlaying);
        }
    }

    private void complete() {
        if (currentPosition >= mediaItemList.size() - 1) {
            currentPosition = 0;
        } else {
            currentPosition++;
        }
        play(currentPosition);
    }
```

**Comment by dragitpro** [comment]
@ojw28 Hello, we are facing same problem and using setMediaSource  (because using MergingMediaSource to merge audio and video streams) instead of setMediaItem. And problem still occurs with:

setKeepContentOnPlayerReset(false);
stop() 
setMediaSource (...) 
prepare(...).
